### PR TITLE
Fix drop/attack animations scaling very badly with pop

### PR DESCRIPTION
**Pull request by TiviPlus**
## About The Pull Request

Believe it or not using GLOB.clients for something that happens very frequently and only needs viewers is a bad idea

On tgmc this was the largest overtime source in the profiler on a 100 pop round and flick overlay was 12th in self cpu

Based on testing with faking 50 clients connected/4 mobs in view this fix was much much faster

## Changelog
:cl:
fix: fixed pickup animations performing worse the more players were connected
/:cl:

### CODE REVIEW

**Review by LemonInTheDark — APPROVED** [review]
wow I'm stupid

**Review by MrMelbert — COMMENTED** [review]
This has some side effects:
- Ais cannot see pickup / attack animations
- You will not able pick up / attack animations of people at the edge of your screen (widescreen)
- Things that extend your view range (scopes and binoculars) will also not see pickup / attack animations
- People on a lower Z-level or upper Z-level will not show attack or pickup animations
- Camera consoles will not show attack or pickup animations

These animations do not *only* need `viewers`, as there are many ways to "view" a mob which does not include you in `viewers`

**Comment by Pumpkinoe** [comment]
seeing attack animations is very important for the AI, I'd say

**Comment by LemonInTheDark** [comment]
> This has some side effects:
> 
>     * Ais cannot see pickup / attack animations

Snowflake for this

>     * You will not able pick up / attack animations of people at the edge of your screen (widescreen)

Fixable with passing args into viewers since it defaults to world.view. someone needs to do a pass on that someday.

>     * Things that extend your view range (scopes and binoculars) will also not see pickup / attack animations

This is really annoying. Solving it depends on where the performance impact of this is, either in the iteration or in client.images setting. Also need to know how much this impacts us in particular to see how big a difference it would make

>     * People on a lower Z-level or upper Z-level will not show attack or pickup animations

Solvable generically. check the vis_locs of the object, and run viewers calls on those as well

>     * Camera consoles will not show attack or pickup animations

snowflake this

**Comment by TiviPlus** [comment]
Im willing to do the AI, z level and fix it for widescreen is that good enough for you

**Comment by TiviPlus** [comment]
Minorly improved flick overlay itself, the viewers in flick to viewers uses widescreen size, includes viewers near vis locs, all camera eyes added as well
Also fixed laser pointer to use widescreen instead of hardcode because I was using it to test

File: code/__HELPERS/game.dm [hunk 1]
```diff
@@ -119,21 +119,39 @@
...
 	var/list/viewing = list()
-	for(var/mob/viewer as anything in viewers(target))
-		if(viewer.client)
+	//get player widescreen screen size
+	var/size = CONFIG_GET(string/default_view)
+	for(var/mob/viewer as anything in viewers(size, target))
+		if(viewer.client) //yes we check this twice but the extra iteration is only worth it at about 50% clientless mobs
 			viewing += viewer.client
+
+	if(!isarea(target))
```
Comment (LemonInTheDark) [inline code/__HELPERS/game.dm:145]: what if we just assert that you cannot show an area to people

File: code/__HELPERS/game.dm [hunk 2]
```diff
@@ -119,21 +119,39 @@
...
+		//can actually be a turf as well because vis locs moment
+		var/atom/movable/vis_locs_access = target
+		for(var/atom/location as anything in vis_locs_access.vis_locs)
+			for(var/mob/viewer as anything in viewers(size, location))
+				if(viewer.client)
+					viewing += viewer.client
+
+	for(var/mob/camera/viewer as anything in GLOB.cliented_mob_cams)
+		viewing += viewer.client
 	flick_overlay(image_to_show, viewing, duration)
```
Comment (LemonInTheDark) [inline code/__HELPERS/game.dm:155]: |= here I think

File: code/__HELPERS/game.dm (hunk 1 above)
Comment (TiviPlus) [reply code/__HELPERS/game.dm:145]: I wouldve straight up used : here so the question is how much do you trust coders to not be stupid

Comment (LemonInTheDark) [reply code/__HELPERS/game.dm:145]: just do a check for it and then CRASH if it passes, it's fine to be explicit this code isn't quite so hot

File: code/modules/admin/verbs/admingame.dm [hunk 3]
```diff
@@ -353,7 +353,7 @@ Traitors and the like can also be revived with the previous role mostly intact.
 	if(view_size.getView() == view_size.default)
 		view_size.setTo(input("Select view range:", "FUCK YE", 7) in list(1,2,3,4,5,6,7,8,9,10,11,12,13,14,37) - 7)
 	else
-		view_size.resetToDefault(getScreenSize(prefs.read_preference(/datum/preference/toggle/widescreen)))
+		view_size.resetToDefault()
```
Comment (LemonInTheDark) [inline code/modules/admin/verbs/admingame.dm:356]: why?

Comment (TiviPlus) [reply code/modules/admin/verbs/admingame.dm:356]: image

File: code/__HELPERS/game.dm (hunk 2 above)
Comment (TiviPlus) [reply code/__HELPERS/game.dm]: done

File: code/modules/admin/verbs/admingame.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/modules/admin/verbs/admingame.dm:356]: I hate old me

**Comment by Mothblocks** [comment]
Conflicts

**Comment by Mothblocks** [comment]
This is still conflicting, please have a maintainer reopen when you fix the conflicts

**Pull request closed without merging**